File: DataDriven/Phoenix/app/lib/app_web/controllers/post_controller.ex
defmodule AppWeb.PostController do
  use AppWeb, :controller
  import Ecto.Query
  alias App.{Repo}
  alias App.Blog.{Post, Comment, Author}

  def create(conn, %{"postData" => post_data}) do
    last_updated_str = Map.get(post_data, "lastUpdated", "")

    last_updated =
      if last_updated_str != "" do
        case NaiveDateTime.from_iso8601("#{last_updated_str} 00:00:00") do
          {:ok, naive_datetime} -> naive_datetime
          _error -> nil
        end
      else
        nil
      end

    post_data = Map.put(post_data, "lastUpdated", last_updated)

    changeset = Post.changeset(%Post{}, post_data)

    authors_ids = Map.get(post_data, "authors", [])

    authors =
      Author
      |> where([a], a.id in ^authors_ids)
      |> Repo.all()

    changeset =
      changeset
      |> Ecto.Changeset.put_assoc(:authors, authors)

    case Repo.insert(changeset) do
      {:ok, _post} ->
        json(conn, %{success: true})

      {:error, changeset} ->
        error_message = "Error creating database entry: #{inspect(changeset)}"
        json(conn, %{error: error_message})
    end
  end

  def get(conn, %{"page" => page, "size" => size}) do
    page = String.to_integer(page || "1")
    size = String.to_integer(size || "10")

    query =
      from p in Post,
        preload: [:authors]

    posts = Repo.paginate(query, page: page, page_size: size)

    if length(posts.entries) < 1 do
      json(conn, %{result: "No posts found."})
    else
      json(conn, %{result: posts.entries})
    end
  end

  def bulk_status_update(conn, %{"postIDs" => post_ids, "newStatus" => new_status}) do
    if is_nil(new_status) or not is_list(post_ids) or length(post_ids) < 1 do
      conn
      |> put_status(:unprocessable_entity)
      |> json(%{error: "Invalid request data."})
    else
      {count, _} =
        from(p in Post, where: p.id in ^post_ids)
        |> Repo.update_all(set: [postStatus: new_status])

      if count > 0 do
        json(conn, %{success: true})
      else
        conn
        |> put_status(:internal_server_error)
        |> json(%{error: "Error updating database entry!"})
      end
    end
  end

  def edit(conn, %{"postData" => post_data}) do
    id = post_data["id"]

    if is_nil(id) or map_size(post_data) < 1 do
      conn
      |> put_status(:unprocessable_entity)
      |> json(%{error: "Invalid request."})
    else
      post = Repo.get!(Post, id)
      changeset = Post.changeset(post, post_data)

      authors_ids = Map.get(post_data, "authors", [])

      if length(authors_ids) > 0 do
        authors =
          Author
          |> where([a], a.id in ^authors_ids)
          |> Repo.all()

        changeset =
          changeset
          |> Ecto.Changeset.put_assoc(:authors, authors)
      end

      case Repo.update(changeset) do
        {:ok, _post} ->
          json(conn, %{success: true})

        {:error, _changeset} ->
          conn
          |> put_status(:internal_server_error)
          |> json(%{error: "Error updating database entry!"})
      end
    end
  end

  def search(conn, %{"search" => search}) do
    if is_nil(search) do
      conn
      |> put_status(:unprocessable_entity)
      |> json(%{error: "Invalid request data."})
    else
      posts =
        from(p in Post,
          where: ilike(p.title, ^"%#{search}%") or ilike(p.content, ^"%#{search}%"),
          preload: [:authors]
        )
        |> Repo.all()

      if length(posts) < 1 do
        json(conn, %{result: "No posts found."})
      else
        json(conn, %{result: posts})
      end
    end
  end

  def sort_by_date(conn, _params) do
    posts =
      from(p in Post, order_by: [desc: p.lastUpdated], preload: [:authors])
      |> Repo.all()

    if length(posts) < 1 do
      json(conn, %{result: "No posts found."})
    else
      json(conn, %{result: posts})
    end
  end

  def popular_posts(conn, _params) do
    max_posts =
      from(c in Comment,
        join: p in Post,
        on: c.post_id == p.id,
        group_by: p.id,
        select: %{post: p.id, count: count(c.id)},
        order_by: [desc: count(c.id)],
        limit: 10
      )
      |> Repo.all()

    post_ids = Enum.map(max_posts, & &1.post)

    posts =
      from(p in Post,
        where: p.id in ^post_ids,
        order_by: fragment("array_position(?, ?)", ^post_ids, p.id)
      )
      |> Repo.all()
      |> Repo.preload([:authors, :categories])

    if length(posts) < 1 do
      json(conn, %{result: "No posts found."})
    else
      json(conn, %{result: posts})
    end
  end
end

File: DataDriven/Laravel/app/Http/Controllers/PostController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use App\Models\Post;
use App\Models\Comment;
use App\Models\Author;
use App\Models\Category;
use Illuminate\Support\Facades\Concurrency;
use Symfony\Component\Console\Output\ConsoleOutput;

class PostController extends Controller
{
    public function create(Request $request)
    {
        $postData = $request->input('postData');
        try {
            $post = Post::create($postData);
            if ($post) {
                return response()->json(['success' => true], 200);
            } else {
                throw new \Exception('Error creating database entry!');
            }
        } catch (\Exception $e) {
            return response()->json(['error' => $e->getMessage()], 500);
        }
    }

    public function get(Request $request)
    {
        $page = $request->query('page', 1);
        $size = $request->query('size', 10);
        try {
            $posts = Post::with(['authors', 'categories'])
                ->skip($size * ($page - 1))
                ->take($size)
                ->get();
            if ($posts->isEmpty()) {
                return response()->json(['result' => 'No posts found.'], 200);
            } else {
                return response()->json(['result' => $posts], 200);
            }
        } catch (\Exception $e) {
            return response()->json(['error' => $e->getMessage()], 500);
        }
    }

    public function bulkStatusUpdate(Request $request)
    {
        $postIDs = $request->input('postIDs');
        $newStatus = $request->input('newStatus');
        if (!$newStatus || !is_array($postIDs) || count($postIDs) < 1) {
            return response()->json(['error' => 'Invalid request data.'], 422);
        }
        try {
            $updated = Post::whereIn('id', $postIDs)->update(['postStatus' => $newStatus]);
            if ($updated) {
                return response()->json(['success' => true], 200);
            } else {
                throw new \Exception('Error updating database entry!');
            }
        } catch (\Exception $e) {
            return response()->json(['error' => $e->getMessage()], 500);
        }
    }

    public function edit(Request $request)
    {
        $postData = $request->input('postData');
        if (empty($postData) || !isset($postData['id'])) {
            return response()->json(['error' => 'Invalid request.'], 422);
        }
        try {
            $post = Post::find($postData['id']);
            if ($post) {
                $post->update($postData);
                return response()->json(['success' => true], 200);
            } else {
                throw new \Exception('Error updating database entry!');
            }
        } catch (\Exception $e) {
            return response()->json(['error' => $e->getMessage()], 500);
        }
    }

    public function search(Request $request)
    {
        $search = $request->query('search');
        if (!$search) {
            return response()->json(['error' => 'Invalid request data.'], 422);
        }
        try {
            $posts = Post::where('title', 'like', "%{$search}%")
                ->orWhere('content', 'like', "%{$search}%")
                ->get();
            if ($posts->isEmpty()) {
                return response()->json(['result' => 'No posts found.'], 200);
            } else {
                return response()->json(['result' => $posts], 200);
            }
        } catch (\Exception $e) {
            return response()->json(['error' => $e->getMessage()], 500);
        }
    }

    public function sortByDate()
    {
        try {
            $posts = Post::orderBy('lastUpdated', 'desc')->get();
            if ($posts->isEmpty()) {
                return response()->json(['result' => 'No posts found.'], 200);
            } else {
                return response()->json(['result' => $posts], 200);
            }
        } catch (\Exception $e) {
            return response()->json(['error' => $e->getMessage()], 500);
        }
    }

    public function popularPosts()
    {
        try {
            $maxPosts = Comment::selectRaw('post, COUNT(*) as count')
                ->groupBy('post')
                ->orderBy('count', 'desc')
                ->limit(10)
                ->pluck('post');
            if ($maxPosts->isEmpty()) {
                return response()->json(['result' => 'No posts found.'], 200);
            } else {
                $posts = Post::whereIn('id', $maxPosts)
                    ->with(['authors', 'categories'])
                    ->get();
                return response()->json(['result' => $posts], 200);
            }
        } catch (\Exception $e) {
            return response()->json(['error' => $e->getMessage()], 500);
        }
    }

    public function newPosts(Request $request)
    {
        set_time_limit(0);

        // Get input data (make sure it's an array)
        $commentsData = $request->input('cats', []);
        if (!is_array($commentsData) || count($commentsData) < 1) {
            return response()->json(['error' => 'Invalid request data.'], 422);
        }

        try {
            // Split the array into two halves
            $halfwayPoint = ceil(count($commentsData) / 2);
            $firstHalf = array_slice($commentsData, 0, $halfwayPoint);
            $secondHalf = array_slice($commentsData, $halfwayPoint);

            // Run first half concurrently
            Concurrency::run(array_map(function ($postData) {
                return function () use ($postData) {
                    // Check if post with the same slug exists, if so, skip
                    if (Post::where('slug', $postData['slug'])->exists()) {
                        return; // Skip this post
                    }

                    // Create the post
                    $post = Post::create($postData);

                    // Output and log the post data
                    $output = new ConsoleOutput();
                    $output->writeln($post);

                    // Log post data
                    log($postData);

                    // Handle authors if present
                    if (isset($postData['authors'])) {
                        $authorEmails = array_column($postData['authors'], 'email');
                        $authorIds = Author::whereIn('email', $authorEmails)->pluck('id')->toArray();
                        $post->authors()->sync($authorIds);
                    }

                    // Handle categories if present
                    if (isset($postData['categories'])) {
                        $categoryNames = array_column($postData['categories'], 'name');
                        $categoryIds = Category::whereIn('name', $categoryNames)->pluck('id')->toArray();
                        $post->categories()->sync($categoryIds);
                    }
                };
            }, $firstHalf)); // Execute concurrently for the first half

            // After the first half finishes, run the second half concurrently
            Concurrency::run(array_map(function ($postData) {
                return function () use ($postData) {
                    // Same logic for the second half of posts
                    if (Post::where('slug', $postData['slug'])->exists()) {
                        return; // Skip this post
                    }

                    $post = Post::create($postData);
                    $output = new ConsoleOutput();
                    $output->writeln($post);
                    log($postData);

                    if (isset($postData['authors'])) {
                        $authorEmails = array_column($postData['authors'], 'email');
                        $authorIds = Author::whereIn('email', $authorEmails)->pluck('id')->toArray();
                        $post->authors()->sync($authorIds);
                    }

                    if (isset($postData['categories'])) {
                        $categoryNames = array_column($postData['categories'], 'name');
                        $categoryIds = Category::whereIn('name', $categoryNames)->pluck('id')->toArray();
                        $post->categories()->sync($categoryIds);
                    }
                };
            }, $secondHalf)); // Execute concurrently for the second half

            return response()->json(['success' => true], 200);
        } catch (\Exception $e) {
            return response()->json(['error' => $e->getMessage()], 500);
        }
    }
}
